[11.0][OU-FIX] #2339, purge obsolete models and fields

StefanRijnhart opened the pull request:
Fixes #2339

* add loaded models and fields to the set of loaded XMLIDs
* purge models and fields with noupdate NULL instead of FALSE

Also

* Fix update_fields_xmlid searching for fields by ir.model.fields instead of
the actual model that the field belongs to
* fix KeyError in case of a missing model in ir.model.fields' unlink
* remove coverage as it is giving out red marks to PRs for no reason
* check reference count before deleting a record after module upgrade
* improve logging when deletion fails (should be rare now)
* reduce logging of deletions in CI
* remove ir.model.relation from ir.model's unlink


This PR is part of a set of PRs for OpenUpgrade 9.0 up to 13.0. Background:

While Odoo deletes obsolete field and model entries from the data model
metadata explicitely in their migration scripts, in OpenUpgrade I have
always meant to rely on the mechanism of purging 'untouched' XMLIDs that
takes care of the deletion of obsolete data records (e.g. views).

However, this mechanism was not applied to field and model entries because
their XMLIDs were created with noupdate NULL instead of FALSE and as such
excluded in the query to gather all obsolete data records.

Also missing was marking the XMLIDs of fields and models as loaded in the
first place.

All of this is working properly in Odoo 13 (introduced gradually across new
releases) so all of this is backported from newer versions one way or
another.

StefanRijnhart committed:
[OU-FIX] #2339, purge obsolete models and fields

Fixes #2339

* add loaded and fields models to the set of loaded XMLIDs
* purge models and fields with noupdate NULL instead of FALSE

Also

* Fix update_fields_xmlid searching for fields by ir.model.fields instead of
the actual model that the field belongs to
* fix KeyError in case of a missing model in ir.model.fields' unlink
* remove coverage as it is giving out red marks to PRs for no reason
* check reference count before deleting a record after module upgrade
* improve logging when deletion fails (should be rare now)
* reduce logging of deletions in CI
* remove ir.model.relation from ir.model's unlink

This PR is part of a set of PRs for OpenUpgrade 9.0 up to 13.0. Background:

While Odoo deletes obsolete field and model entries from the data model
metadata explicitely in their migration scripts, in OpenUpgrade I have
always meant to rely on the mechanism of purging 'untouched' XMLIDs that
takes care of the deletion of obsolete data records (e.g. views).

However, this mechanism was not applied to field and model entries because
their XMLIDs were created with noupdate NULL instead of FALSE and as such
excluded in the query to gather all obsolete data records.

Also missing was marking the XMLIDs of fields and models as loaded in the
first place.

All of this is working properly in Odoo 13 (introduced gradually across new
releases) so all of this is backported from newer versions one way or
another.

StefanRijnhart commented on odoo/addons/base/ir/ir_model.py:
```diff
@@ -829,7 +841,7 @@ def _reflect_field(self, field):
             "FROM ir_model_fields f LEFT JOIN ir_model_data d "
             "ON f.id=d.res_id and d.model='ir.model.fields' WHERE f.model=%s",
-            (self._name,))
+            (field.model_name,))
```
@pedrobaeza I think this is the source of the problems you observed with fields not being moved to their new home module properly. My mistake when porting the framework to 11.0. Sorry for the inconvenience!

pedrobaeza replied:
No worries. The important thing is that it gets fixed in the best possible way.

StefanRijnhart commented:
Observed data loss so I'm investigating.

StefanRijnhart committed:
[OU-FIX] Check the model of the field, not ir.model.fields

After Odoo 10, this code had to be moved from ir.model (the actual model
of the field) to ir.model.fields. I must have been sleeping when porting it.

StefanRijnhart commented:
The observed data loss was a dropped field definition of res.partner::property_stock_customer and the related ir.property entries, amongst similar problems. The problem was due to a number of additional problems due to incorrect porting of this method after it was moved from the field's related model to ir.model.fields itself in Odoo 11.0. Fixed in the additional commit. The problem was specific to OpenUpgrade 11.0 and does not occur in the other branches related to #2339.
